test/gtest/gpu_nchw_nhwc_transpose.cpp

```diff
@@ -0,0 +1,658 @@
+{
+    cpu_ncdhw2ndhwc<T>(dst, src, N, C, 1, H, W);
+}
+
+template <typename T>
+void cpu_ndhwc2ncdhw(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+{
+    for(uint64_t i_n = 0; i_n < N; i_n++)
+    {
+        for(uint64_t i_c = 0; i_c < C; i_c++)
+        {
+            for(uint64_t i_d = 0; i_d < D; i_d++)
+            {
+                for(uint64_t i_h = 0; i_h < H; i_h++)
+                {
+                    for(uint64_t i_w = 0; i_w < W; i_w++)
+                    {
+                        uint64_t idx_ndhwc =
+                            i_n * D * H * W * C + i_d * H * W * C + i_h * W * C + i_w * C + i_c;
+                        uint64_t idx_ncdhw =
+                            i_n * C * D * H * W + i_c * D * H * W + i_d * H * W + i_h * W + i_w;
+                        dst[idx_ncdhw] = src[idx_ndhwc];
+                    }
+                }
+            }
+        }
+    }
+}
+
+template <typename T>
+void cpu_nhwc2nchw(T* dst, T* src, uint64_t N, uint64_t C, uint64_t H, uint64_t W)
+{
+    cpu_ndhwc2ncdhw<T>(dst, src, N, C, 1, H, W);
+}
+
+template <typename T, typename TRANSPOSE_SOL>
+struct cpu_transpose
+{
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionDefault2Nhwc>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t H, uint64_t W)
+    {
+        cpu_nchw2nhwc<T>(dst, src, N, C, H, W);
+    }
+
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ncdhw2ndhwc<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionDefault2Ndhwc>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ncdhw2ndhwc<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionNhwc2Default>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t H, uint64_t W)
+    {
+        cpu_nhwc2nchw<T>(dst, src, N, C, H, W);
+    }
+
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ndhwc2ncdhw<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionNdhwc2Default>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ndhwc2ncdhw<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename TRANSPOSE_SOL>
+struct transpose_str
+{
+};
+
+template <>
+struct transpose_str<miopen::TransposeSolutionDefault2Nhwc>
+{
+    static std::string get() { return "nchw2nhwc"; }
+};
+
+template <>
+struct transpose_str<miopen::TransposeSolutionNhwc2Default>
+{
+    static std::string get() { return "nhwc2nchw"; }
+};
+
+template <typename T>
+struct to_miopen_data_type
+{
+};
+
+template <>
+struct to_miopen_data_type<float>
```

Comment:
why do you need this? It's already there in `tensor_holder.hpp`.

Reply:
It comes from the original file, but agree, it should be replaced with `struct miopen_type;` from `tensor_holder.hpp`

Reply:
Very nice, yeah, I don't know the original author's motivation for those, but `miopen_type` was an easy drop-in replacement.